PP-587: One MoM reports the compute node info to server

lisa-altair opened the pull request:
This introduces the vnode_pool attribute.
Server and mom improvements were also made.

<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-587](https://pbspro.atlassian.net/browse/PP-587)**

#### Problem description
* On a Cray X*-series system, each MoM will report the same underlying set of compute nodes.  This can take up a lot of of the server's time, as well as take up communication bandwidth.

#### Cause / Analysis
* The server-MoM communication works such that the server asks each MoM to report about herself and any of vnodes she knows about.  On a Cray X*-series this is unnecessary since the MoMs are all reporting the same set of underlying compute nodes. 

#### Solution description
* Added a new attribute vnode_pool which allows the admin to specify which underlying pool of compute nodes is being reported by each mom.  Therefore all of the moms that belong to the same vnode_pool are reporting/responsible for the same set of compute nodes.  In this way the server only has to ask one mom for the compute node information, and does not have to process the same information multiple times.

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

arungrover commented on src/server/node_manager.c:
```diff
@@ -7778,6 +7999,7 @@ req_momrestart(struct batch_request *preq)
 {
 	mominfo_t *pmom;
 	int        stm;
+	int	   com;
```
This variable is not needed.

arungrover commented on src/server/req_manager.c:
```diff
@@ -178,6 +188,16 @@ extern int delete_attr_db(pbs_db_conn_t *conn,
+ * to do a faster lookup of hostnames.
+ * It is stored against pname in make_host_addresses_list()
+ */
+struct pul_store {
```
I couldn't understand the significance of the work "pul" can we rename it?

lisa-altair replied:
Sure, but I can't figure out what to name it.  Other things in the PBS code are already too similar (but for different purposes).  Can you provide a suggestion?

lisa-altair replied:
I should note that I'm following existing code...if I had to guess it stands for pointer to unsigned long.

arungrover replied:
Since we now know what the structure name means, I'm okay to have it as is. Thanks

arungrover commented on src/server/req_manager.c:
```diff
@@ -1766,6 +1786,191 @@ mgr_queue_unset(struct batch_request *preq)
+	/* pool remove her from that role and if another Mom in the pool and */
+	/* is up, make that Mom the new inventory Mom */
+
+	if (psvrmom->msr_vnode_pool != 0) {
```
There are many places where we try to find whether or not a mom belongs to a vnode_pool. Can we make this a function?

lisa-altair replied:
I see what you mean.   Sure, I'll make that part a function.

arungrover commented on src/server/req_manager.c:
```diff
@@ -2449,10 +2673,169 @@ make_host_addresses_list(char *phost, u_long **pul)
+ * @par MT-safe: No
+ **/
+AVL_IX_DESC *
+cr_tree(int dups)
```
These functions are generic in nature and they can go to avltree.c

lisa-altair replied:
avltree.c covers the index routines, and was written by Gregory Tseytin.  I am hesitant to add to it...should I?

arungrover commented on src/server/req_manager.c:
```diff
@@ -3149,13 +3554,41 @@ struct batch_request *preq;
+				ppool = ppool->vnpm_next;
+			}
+			if (ppoolm) {
+				sprintf(log_buffer, "POOL: cross linking %d vnodes from %s",
```
This log message is also not covered in the design proposal. Please add it.

lisa-altair replied:
Added as interface 11.

arungrover commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,337 @@
+                         1, msg="an inventory mom not chosen correctly")
+
+        # Only one mom is inventory mom
+        if (found_in_momA == 0):
```
variable "found_in_momA" is set when it gets hello_no_inventory, so we can't mark that mom as inventory mom.

vccardenas replied:
This looks fine to me. At this point either "found_in_momA" or "found_in_momB" could have a value of 0 because we are not guaranteed which mom will be the inventory mom.  After line 154 we know there is only one inventory mom.  If "found_in_momA" is set to 0 then it did not get hello_no_inventory which makes momA the inventory mom.  If "found_in_momA" is not set to 1 then it got hello_no_inventory which makes momB the inventory mom.

arungrover replied:
You are right! I interpreted it wrongly

arungrover commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,337 @@
+        rc = self.server.log_match(
+            "Setting inventory_mom for vnode_pool %s to %s" %
+            ('1', self.noninv_mom.shortname), max_attempts=5,
+            starttime=start_time)
```
we should change start_time again just before killing the inventory mom.

lisa-altair replied:
Agreed.  Fixed.

vccardenas commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,337 @@
+        # Start the previous inv mom
+        self.inv_mom.start()
+
+        # Check previous inventory mom is up and no longer the inventory mom
```
The EDD mentions "If the Inventory Mom goes down, then another Mom from the same vnode_pool is chosen to be the Inventory Mom." but does not say that once the previous inventory mom is back up that it will no longer be the inventory mom.  The doc string and the comment in line 216 needs updating too.  Starting the previous inventory mom might be useful only for the purpose of having two moms up which might be needed by the other tests.

lisa-altair replied:
Fixed.

lisa-altair commented on src/server/node_manager.c:
```diff
@@ -224,6 +226,8 @@ extern int parse_prov_vnode(char *,exec_vnode_listtype *);
 static void check_and_set_multivnode(struct pbsnode *);
 static void propagate_socket_licensing(mominfo_t *);
+int write_single_node_mom_attr(struct pbsnode *np);
+char *msg_new_inventory_mom = "Setting inventory_mom for vnode_pool %d to %s";
```
ok.

lisa-altair commented on src/server/node_manager.c:
```diff
@@ -1015,6 +1027,47 @@ momptr_down(mominfo_t *pmom, char *why)
+		if (ret != DIS_SUCCESS)
+			return (ret);
+	}
+	rpp_flush(stream);
```
Ok.

lisa-altair commented on src/resmom/mom_main.c:
```diff
@@ -9451,7 +9451,7 @@ main(int argc, char *argv[])
 
 		if (internal_state_update) {
-			state_to_server();
+			state_to_server(0);
```
Created UPDATE_VNODES and UPDATE_MOM_ONLY

lisa-altair commented on src/server/node_func.c:
```diff
@@ -869,7 +856,7 @@ static char *nodeerrtxt = "Node description file update failed";
 /**
  * @brief
- *		Static function to update all the nodes in the db, which have
+ *		Static function to update the specified mom in the db, which have
```
Fixed the comment.

lisa-altair commented:
I have addressed the comments.  And I did add the AVL tree code to avltree.c.  I also discovered that there were TPP functions for the same AVL tree functions I was adding.  So I made the functions generic and changed the TPP code and the server code to call the generic functions.  Please have a look and let me know your thoughts.

arungrover commented on src/server/req_manager.c:
```diff
@@ -2689,15 +2699,15 @@ make_host_addresses_list(char *phost, u_long **pul)
 
 		if (hostaddr_tree == NULL ) {
-			hostaddr_tree = cr_tree(0);
+			hostaddr_tree = create_tree(0, 0);
```
Please replace the first argument with AVL_NO_DUP_KEYS. This is needed in other create_tree instances too (wherever we don't need duplicates in the tree)

lisa-altair replied:
Good point.

lisa-altair commented on src/lib/Libtpp/tpp_router.c:
```diff
@@ -1325,7 +1325,7 @@ router_pkt_handler(int tfd, void *data, int len, void *c)
 				}
 
-				if (tpp_tree_add_del(r->AVL_my_leaves, &l->leaf_addrs[0], l, TPP_OP_ADD) != 0) {
+				if (tree_add_del(r->AVL_my_leaves, &l->leaf_addrs[0], l, TPP_OP_ADD) != 0) {
```
Oh, good catch.  Looks like I missed some.

lisa-altair commented on src/server/node_func.c:
```diff
@@ -2339,11 +2340,12 @@ decode_Mom_list(struct attribute *patr, char *name, char *rescn, char *val)
 			 */
 			if (momfqdn_tree == NULL)
-				momfqdn_tree = cr_tree(0);
+				momfqdn_tree = create_tree(0, 0);
```
Good point.  Will do.

lisa-altair commented on src/lib/Libutil/avltree.c:
```diff
@@ -130,6 +130,8 @@ typedef struct {
 static pthread_once_t avl_init_once = PTHREAD_ONCE_INIT;
 static pthread_key_t avl_tls_key;
 
+#define MAX_AVLKEY_LEN 1024
```
I think you're right.  I took a look at the new usages I'm adding, and I *think* a length of 100 should be enough.

lisa-altair commented:
I have made the changes, please take a look.

arungrover commented on src/server/req_manager.c:
```diff
@@ -2521,49 +2783,58 @@ create_pbs_node2(char *objname, svrattrl *plist, int perms, int *bad, struct pbs
+
+		/* create node tree if not already done */
+		if (node_tree == NULL ) {
+			node_tree = create_tree(0, 0);
```
Please add "AVL_NO_DUP_KEYS" in this call as well.

lisa-altair replied:
done.  I checked for other instances, and I think we've found them all now.  Thanks.

arungrover commented on src/resmom/mom_server.c:
```diff
@@ -1462,7 +1477,7 @@ state_to_server(void)
 	DBPRT(("updating state 0x%x to server\n", i))
 
-	if (vnlp != NULL)
+	if ((vnlp != NULL) && (mom_only == 0))
```
Instead of comparing mom_only with 0, please use UPDATE_VNODES. 
Suggestion - may be change the variable name itself from "mom_only" to "update_what"

lisa-altair replied:
That makes sense, I've made some changes.  Please take a look.

lisa-altair commented:
@arungrover Thanks for the review.  I have made the changes you requested.

arungrover commented:
@lisa-altair change looks good. Let's wait for another reviewer to review it.
Thanks!

vccardenas commented:
The PTL tests look good to me.

arungrover commented:
@lisa-altair Please squash and rebase for merge

bhroam commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,325 @@
+            self.assertEqual(rc, 0)
+        except PbsStatusError as e:
+            self.assertEquals(
+                e.msg[0], "/opt/pbs/bin/pbsnodes: Server has no node list")
```
I have two comments.  First, this only works if the test is run in CLI mode.  This message won't be printed if it is run in API mode.  Second, what happens if pbsnodes is in a different location than /opt/pbs/bin?  There's nothing saying that PBS has to be installed there.

vccardenas replied:
Good points - sounds like this can be added to "best practices".  Updated this part as well as similar areas in the other tests.

bhroam commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,325 @@
+                                attrib=attr_A, expect=False)
+        except PbsManagerError as e:
+            self.assertEquals(
+                e.msg[1], "qmgr: Error (15014) returned from server")
```
This message will only be printed in CLI mode.  You should make this test more generic so it can be run in either mode (same for the rest of these in the file).

vccardenas replied:
Updated this part as well as similar areas in the other tests.

bhroam commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,325 @@
+            self.assertEquals(
+                e.msg[1], "qmgr: Error (15014) returned from server")
+
+        attr_A = {'vnode_pool': '0'}
```
Just curious, why is this called attr_A when you use just attr down below?  It confused me a little.

vccardenas replied:
It refers to attributes for momA.  I see attr_A instead of attr used down below.

bhroam commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,325 @@
+        attr_A = {'vnode_pool': '0'}
+        try:
+            self.server.manager(MGR_CMD_CREATE, NODE, id=self.hostA,
+                                attrib=attr_A, expect=False)
```
expect=False is the default, you don't need to pass it if you don't want to

vccardenas replied:
Taken out expect=False, since it is the default, here and in other areas where it is used.

bhroam commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,325 @@
+        try:
+            self.momB = self.moms.values()[1]
+        except IndexError:
+            self.assertTrue(
```
This will fail the test if there aren't 2 moms.  Do an if and skip_test()

vccardenas replied:
The test requires two moms and if it were to silently skip the test then we would have no failure or error to indicate that the test used less than two moms.

vccardenas replied:
I reconsidered your comment.  I agree that the tests should be skipped if the required test environment is not there.

bhroam commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,325 @@
+            self.noninv_mom = self.momB
+        else:
+            self.inv_mom = self.momB
+            self.noninv_mom = self.momA
```
Why make these class variables?  They can't be used between tests because you don't know the order in which the tests will run.  Why not just make them locals?

vccardenas replied:
These were changed to locals.

bhroam commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,325 @@
+            self.logger.info(
+                "Inventory mom %s is not in server logs." %
+                (self.noninv_mom.shortname))
+            self.assertTrue(rc)
```
I'd not use the if.  I'd use the msg parameter to self.assertTrue().  You can follow the it with a self.logger of the following message.  The second self.assertTrue() is not needed.

vccardenas replied:
Good point.  Done.

bhroam commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,325 @@
+                VNODE, {'resources_available.vntype': 'cray_compute'})
+            vlist = vnl["resources_available.vntype=cray_compute"]
+        except Exception:
+            sys.exc_clear()
```
I don't think this is required.  I think you want to say 'pass' instead

vccardenas replied:
Done.

bhroam commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,325 @@
+
+        # Loop through each compute vnode in the list and check if state = free
+        r1 = len(vlist)
+        for i in range(0, r1):
```
This would be more pythony if you changed it to for v1 in vlist:

vccardenas replied:
Range is in python too isn't it.  This was changed to be more pythony.

bhroam replied:
Yes range() is part of python and has its place.  I use it when I need the index outside of just indexing into a list.  I find it more C like if you use it to just index into a list.  If all you are doing is iterating over the values in a list, I find it more pythony to do "for v in list:".  Thanks for making the change.

bhroam commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,325 @@
+        try:
+            self.momB = self.moms.values()[1]
+        except IndexError:
+            self.assertTrue(
```
change this to skip_test()

vccardenas replied:
The test requires two moms and if it were to silently skip the test then we would have no failure or error to indicate that the test used less than two moms.

bhroam commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,325 @@
+            found_in_momB = 1
+
+        if (found_in_momA + found_in_momB != 0):
+            self.assertTrue(False, msg="Both moms must report inventory")
```
Put the conditional from your if into the self.assertTrue().  I'm confused by your conditional.  Won't this fail the test unless the conditional == 0?  That seems wrong to me.  Isn't the success case conditional == 2?

vccardenas replied:
This has been clarified per your suggestion.  The desired is conditional == 0 which indicates that both moms are inventory moms (i.e. both moms did not have the "no inventory" message in their logs).

bhroam commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,325 @@
+        rv = self.server.log_match(
+            "Unsupported actions for vnode_pool", max_attempts=5,
+            starttime=start_time)
+        self.assertTrue(rv, msg="Found correct server log message")
```
the msg will only print if the assert fails.  The message sounds like it passed.

vccardenas replied:
The "msg=..." has been taken out and placed in a logger.info().

vccardenas commented:
In the interest of time I will be making changes to the PTL based on Bhroam's comments.  I'll also address each comment.

hirenvadalia commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,312 @@
+    """
+
+    def setUp(self):
+        TestFunctional.setUp(self)
```
Please add a check to verify that testsuite has two moms.

vccardenas replied:
Done.  I have removed this check from tests now that it is checked in setUp().

hirenvadalia commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,312 @@
+        # Clean all default nodes because each test case will set up nodes.
+        try:
+            self.server.status(NODE)
+            rc = self.server.manager(MGR_CMD_DELETE, NODE, None, "")
```
I think this self.server.manager should be outside of try and except block.
Also you don't need 'rc' as if delete fails then manager() will throw an PbsManagerError.

vccardenas replied:
It should remain in the try block because we only want to delete the nodes if they exist.  There will be an error from server when deleting nodes that don't exist.

hirenvadalia commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -0,0 +1,312 @@
+        start_time = int(time.time())
+
+        # Kill inventory mom
+        inv_mom.stop(sig='-KILL')
```
I would suggest to use inv_mom.signal() here instead .stop().

vccardenas replied:
Both send a signal (in this case KILL) to the mom's pid.

hirenvadalia replied:
Yes, that's implementation part. But by design signal() is for sending signal to mom (which you are doing in test case) and stop() is for stopping mom. I would still suggest to use .signal() to send KILL command, rest up to you.

vccardenas replied:
@hirenvadalia, I have changed to signal('KILL') per your suggestion.

hirenvadalia commented on test/tests/functional/pbs_cray_vnode_pool.py:
```diff
@@ -49,6 +49,11 @@ class TestVnodePool(TestFunctional):
 
     def setUp(self):
         TestFunctional.setUp(self)
+        try:
```
rather than try and except you can use assertEqual(len(self.moms.values()), 2)

vccardenas replied:
@hirenvadalia, I have made this change and added a msg.

vccardenas replied:
@hirenvadalia, I reconsidered this and the tests should be skipped if the required test environment is not there.

bhroam commented:
The PTL test looks good to me.  Thanks for making the changes.

arungrover commented:
@lisa-altair  please squash and rebase the branch.

lisa-altair commented:
I have squashed and rebased.  Please take it forward.

arungrover commented:
Merged
